blueman/main/DbusService.py

```diff
@@ -1,56 +1,132 @@
-        setattr(self.__class__, method.__name__, wrapper)
-
-    def remove_definitions(self, obj):
-        for name, func in self._definitions(obj):
-            delattr(self.__class__, name)
-
-        self.__class__._refresh_dbus_registration()
-
-    @classmethod
-    def _refresh_dbus_registration(cls):
-        cls.__class__.__init__(cls, cls.__name__, cls.__bases__, cls.__dict__)
+import logging
+import traceback
+
+from gi.repository import Gio, GLib
+
+
+class DbusError(Exception):
+    _name = "org.blueman.Error"
+
+    def __init__(self, message=None):
+        self._message = message
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def message(self):
+        return self._message
+
+
+class DbusService:
+    def __init__(self, bus_name, interface_name, path, bus_type):
+        self._bus = Gio.bus_get_sync(bus_type)
+        if bus_name:
+            Gio.bus_own_name(bus_type, bus_name, Gio.BusNameOwnerFlags.NONE, None, None, None)
+        self._methods = {}
+        self._signals = {}
+        self._interface_name = interface_name
+        self._path = path
+        self._regid = None
+
+    def add_method(self, name, arguments, return_value, method, pass_sender=False, is_async=False):
+        options = set()
+        if pass_sender:
+            options.add("sender")
+        if is_async:
+            options.add("async")
+        self._methods[name] = (arguments, return_value, method, options)
+        self._reregister()
+
+    def remove_method(self, name):
+        del self._methods[name]
+        self._reregister()
+
+    def add_signal(self, name, signature):
+        self._signals[name] = signature
+        self._reregister()
+
+    def remove_signal(self, name):
+        del self._signals[name]
+        self._reregister()
+
+    def emit_signal(self, name, *args):
+        self._bus.emit_signal(None, self._path, self._interface_name, name,
+                              self._prepare_arguments(self._signals[name], args))
+
+    def register(self):
+        node_xml = "<node name='/'><interface name='%s'>" % self._interface_name
+        for method_name, method_info in self._methods.items():
+            node_xml += "<method name='%s'>" % method_name
+            for argument in method_info[0]:
+                node_xml += "<arg type='%s' direction='in'/>" % argument
+            if method_info[1]:
+                node_xml += "<arg type='%s' direction='out'/>" % method_info[1]
+            node_xml += "</method>"
+        for signal_name, signal_signature in self._signals.items():
+            node_xml += "<signal name='%s'>" % signal_name
+            if signal_signature:
+                node_xml += "<arg type='%s'/>" % signal_signature
+            node_xml += "</signal>"
+        node_xml += "</interface></node>"
+
+        node_info = Gio.DBusNodeInfo.new_for_xml(node_xml)
+
+        regid = self._bus.register_object(
+            self._path,
+            node_info.interfaces[0],
```

Comment:
Please add comment that this assumes only a single interface.

Reply:
Should be obvious from the method signatures as the constructor takes a (single) interface name while `add_method` and `add_signal` don't, no?

Reply:
It is now but will it in a year or two? Also it is very easy to handle it already but I'm not going to hold the PR for only this.